source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,405 @@ namespace EnvoyDefault {
+HeaderValidator::validateStatusHeader(const HeaderString& value) {
+  // Validate that the response :status header is a valid whole number between 100 and 999
+  // (inclusive). This is based on RFC 9110, although the Envoy implementation is more permissive
+  // and allows status codes larger than 599,
+  // https://www.rfc-editor.org/rfc/rfc9110.html#section-15:
+  //
+  // The status code of a response is a three-digit integer code that describes the result of the
+  // request and the semantics of the response, including whether the request was successful and
+  // what content is enclosed (if any). All valid status codes are within the range of 100 to 599,
+  // inclusive.
+
+  static uint32_t kMinimumResponseStatusCode = 100;
+  static uint32_t kMaximumResponseStatusCode = 999;
+  const auto& value_string_view = value.getStringView();
+
+  // Convert the status to an integer.
+  std::uint32_t status_value{};
+  auto result = std::from_chars(value_string_view.begin(), value_string_view.end(), status_value);
+  if (result.ec != std::errc() || result.ptr != value_string_view.end()) {
+    return {HeaderValueValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidStatus};
+  }
+
+  if (status_value < kMinimumResponseStatusCode || status_value > kMaximumResponseStatusCode) {
+    return {HeaderValueValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidStatus};
+  }
+
+  return HeaderValueValidationResult::success();
 }
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateResponseHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateGenericHeaderName(const HeaderString& name) {
+  // Verify that the header name is valid. This also honors the underscore in
+  // header configuration setting.
+  //
+  // From RFC 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-5.1:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-name     = token
+  // token          = 1*tchar
+  //
+  // tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
+  //                / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
+  //                / DIGIT / ALPHA
+  //                ; any VCHAR, except delimiters
+  const auto& key_string_view = name.getStringView();
+  bool allow_underscores =
+      config_.headers_with_underscores_action() == HeaderValidatorConfig::ALLOW;
+  // This header name is initially invalid if the name is empty.
+  if (key_string_view.empty()) {
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  bool is_valid = true;
+  char c = '\0';
+
+  for (auto iter = key_string_view.begin(); iter != key_string_view.end() && is_valid; ++iter) {
+    c = *iter;
+    is_valid &= testChar(kGenericHeaderNameCharTable, c) && (c != '_' || allow_underscores);
+  }
+
+  if (!is_valid && c == '_' &&
```

Comment:
Just calling out I think this is a departure from prior behavior.
I believe that before, the codecs would validate character validity and Envoy would only drop valid headers with underscores.   I'd be inclined to keep prior behavior but no strong feelings given bad data won't get through either way.

Reply:
I'm slightly confused, can you clarify? From my understanding:

- [`http1/codec_impl.cc:1322`](https://github.com/envoyproxy/envoy/blob/main/source/common/http/http1/codec_impl.cc#L1322) checks each header for an underscore and either drops the header or rejects the request based on the config.
- http_parser is built with permissive parsing when UHV is enabled, so character validation (including checking for underscores) occurs in UHV.

I believe H1 UHV is mirroring this behavior, with the added checks that would typically be performed by http_parser.

Reply:
sorry if I wasn't clear.
In the non-UHV implementation, I believe http parser would first do checks for invalid characters.  For headers which had valid characters, they'd be passed to the HCM which would do underscore checks and drop headers with underscores if configured.  So I believe (and we can test) prior behavior for _invalid_characters would be to reject the headers (at http parser) rather than drop.  If you don't want to tackle it now we could add a TODO to regression test this particular case to  either change to prior behavior or document the difference  (or just have a test that there's no behavioral change if I'm remembering wrong)

Reply:
Gotcha, that makes sense. I'll adjust this check to mirror the current non-UHV behavior.